sql/parser2/eval.go

```diff
@@ -20,31 +20,308 @@ package parser2
 import (
 	"fmt"
 	"math"
+	"reflect"
 	"strconv"
-
-	"github.com/cockroachdb/cockroach/sql/sqlwire"
 )
 
-type opType int
+// TODO(pmattis):
+//
+// - Support tuples (i.e. []Datum) and tuple operations (a IN (b, c, d)).
+//
+// - Support decimal arithmetic.
+
+// A Datum holds either a bool, int64, float64, string or []Datum.
+type Datum interface {
+	IsNull() bool
+	ToBool() (Datum, error)
+	String() string
+}
+
+var _ Datum = dbool(false)
+var _ Datum = dint(0)
+var _ Datum = dfloat(0)
+var _ Datum = dstring("")
+var _ Datum = dnull{}
+
+type dbool bool
+
+func (d dbool) IsNull() bool {
+	return false
+}
+
+func (d dbool) ToBool() (Datum, error) {
+	return d, nil
+}
+
+func (d dbool) String() string {
+	if d {
+		return "true"
+	}
+	return "false"
+}
+
+type dint int64
+
+func (d dint) IsNull() bool {
+	return false
+}
+
+func (d dint) ToBool() (Datum, error) {
+	return dbool(d != 0), nil
+}
+
+func (d dint) String() string {
+	return strconv.FormatInt(int64(d), 10)
+}
+
+type dfloat float64
 
-const (
-	intOp opType = iota
-	uintOp
-	floatOp
-	stringOp
+func (d dfloat) IsNull() bool {
+	return false
+}
+
+func (d dfloat) ToBool() (Datum, error) {
+	return dbool(d != 0), nil
+}
+
+func (d dfloat) String() string {
+	return strconv.FormatFloat(float64(d), 'g', -1, 64)
+}
+
+type dstring string
+
+func (d dstring) IsNull() bool {
+	return false
+}
+
+func (d dstring) ToBool() (Datum, error) {
+	return null, fmt.Errorf("cannot convert string to bool")
+}
+
+func (d dstring) String() string {
+	return string(d)
+}
+
+type dnull struct{}
+
+func (d dnull) IsNull() bool {
+	return true
+}
+
+func (d dnull) ToBool() (Datum, error) {
+	return d, fmt.Errorf("cannot convert NULL to bool")
+}
+
+func (d dnull) String() string {
+	return "NULL"
+}
+
+var null = dnull{}
+
+var (
+	boolType   = reflect.TypeOf(dbool(false))
+	intType    = reflect.TypeOf(dint(0))
+	floatType  = reflect.TypeOf(dfloat(0))
+	stringType = reflect.TypeOf(dstring(""))
+	nullType   = reflect.TypeOf(null)
 )
 
-var null = sqlwire.Datum{}
+type unaryArgs struct {
+	op      UnaryOp
+	argType reflect.Type
+}
+
+// unaryOps contains the unary operations indexed by operation type and
+// argument type.
+var unaryOps = map[unaryArgs]func(Datum) (Datum, error){
+	unaryArgs{UnaryPlus, intType}: func(d Datum) (Datum, error) {
+		return d, nil
+	},
+	unaryArgs{UnaryPlus, floatType}: func(d Datum) (Datum, error) {
+		return d, nil
+	},
+
+	unaryArgs{UnaryMinus, intType}: func(d Datum) (Datum, error) {
+		return -d.(dint), nil
+	},
+	unaryArgs{UnaryMinus, floatType}: func(d Datum) (Datum, error) {
+		return -d.(dfloat), nil
+	},
+
+	unaryArgs{UnaryComplement, intType}: func(d Datum) (Datum, error) {
+		return ^d.(dint), nil
+	},
+}
+
+type binArgs struct {
+	op        BinaryOp
+	leftType  reflect.Type
+	rightType reflect.Type
+}
+
+// binOps contains the binary operations indexed by operation type and argument
+// types.
+var binOps = map[binArgs]func(Datum, Datum) (Datum, error){
```

Comment:
can you extract the function signature into a local type? I think it'd make these easier to read and review, especially since none of them seems to return non-nil errors